File: contracts/contracts/factory.exs
@version 1

export fun(get_pool_code(token1_address, token2_address, pool_address, lp_token_address)) do
  code = ""

  token1_address = String.to_uppercase(token1_address)
  token2_address = String.to_uppercase(token2_address)
  pool_address = String.to_uppercase(pool_address)
  lp_token_address = String.to_uppercase(lp_token_address)

  router_address = @ROUTER_ADDRESS
  factory_address = @FACTORY_ADDRESS
  master_address = @MASTER_ADDRESS
  protocol_fee_address = @PROTOCOL_FEE_ADDRESS

  if token1_address != token2_address do
    if token1_address > token2_address do
      temp = token1_address
      token1_address = token2_address
      token2_address = temp
    end

    code = """
    @POOL_CODE
    """
  end

  code
end

export fun(get_farm_code(lp_token, start_date, end_date, reward_token, farm_genesis_address)) do
  lp_token = String.to_hex(lp_token)
  reward_token = String.to_uppercase(reward_token)

  router_address = @ROUTER_ADDRESS
  factory_address = @FACTORY_ADDRESS
  master_address = @MASTER_ADDRESS

  """
  @FARM_CODE
  """
end

export fun(get_farm_lock_code(lp_token, start_date, end_date, reward_token, farm_genesis_address)) do
  lp_token = String.to_hex(lp_token)
  reward_token = String.to_uppercase(reward_token)

  router_address = @ROUTER_ADDRESS
  factory_address = @FACTORY_ADDRESS
  master_address = @MASTER_ADDRESS

  """
  @FARM_LOCK_CODE
  """
end

export fun(get_lp_token_definition(token1_address, token2_address)) do
  token1_address = String.to_uppercase(token1_address)
  token2_address = String.to_uppercase(token2_address)

  if token1_address > token2_address do
    temp = token1_address
    token1_address = token2_address
    token2_address = temp
  end

  Json.to_string(
    aeip: [2, 8, 18, 19],
    supply: 10,
    type: "fungible",
    symbol: "aeSwapLP",
    name: "aeSwap LP Token",
    allow_mint: true,
    properties: [
      token1_address: token1_address,
      token2_address: token2_address
    ],
    recipients: [
      [
        to: Chain.get_burn_address(),
        amount: 10
      ]
    ]
  )
end

File: lib/application/factory.dart
/// SPDX-License-Identifier: AGPL-3.0-or-later
import 'dart:async';
import 'package:aedex/domain/models/failures.dart';
import 'package:aedex/domain/models/result.dart';
import 'package:aedex/domain/models/util/model_parser.dart';
import 'package:archethic_lib_dart/archethic_lib_dart.dart';

class Factory with ModelParser {
  Factory(this.factoryAddress, this.apiService);

  final String factoryAddress;
  final ApiService apiService;

  /// Return the code to create a pool for a pair of tokens.
  /// [token1Address] is the first token address of the pair
  /// [token2Address] is the second token address of the pair
  /// [poolAddress] is the genesis address of the pool chain
  /// [lpTokenAddress] is the address of the lp token (it should be the creation address of the pool)
  Future<Result<String, Failure>> getPoolCode(
    String token1Address,
    String token2Address,
    String poolAddress,
    String lpTokenAddress,
  ) async {
    return Result.guard(
      () async {
        final result = await apiService.callSCFunction(
          jsonRPCRequest: SCCallFunctionRequest(
            method: 'contract_fun',
            params: SCCallFunctionParams(
              contract: factoryAddress.toUpperCase(),
              function: 'get_pool_code',
              args: [
                token1Address,
                token2Address,
                poolAddress,
                lpTokenAddress,
              ],
            ),
          ),
        );
        return result.toString();
      },
    );
  }

  /// Return the code to create a farm for a LP token and a reward token.
  /// [lpTokenAddress] is the lp token to provide in the farm
  /// [startDate] is the timestamp (in sec) where the farm starts to give reward (should be between 2 hours and 1 week from farm creation date)
  /// [endDate] is the timestamp (in sec) where the farm ends to give reward (should be between 1 month and 1 year from start date)
  /// [rewardTokenAddress] is the reward token address or "UCO"
  /// [farmGenesisAddress] is the genesis address of the farm contract chain
  Future<Result<String, Failure>> getFarmCode(
    String lpTokenAddress,
    int startDate,
    int endDate,
    String rewardTokenAddress,
    String farmGenesisAddress,
  ) async {
    return Result.guard(
      () async {
        final result = await apiService.callSCFunction(
          jsonRPCRequest: SCCallFunctionRequest(
            method: 'contract_fun',
            params: SCCallFunctionParams(
              contract: factoryAddress.toUpperCase(),
              function: 'get_farm_code',
              args: [
                lpTokenAddress,
                startDate,
                endDate,
                rewardTokenAddress,
                farmGenesisAddress,
              ],
            ),
          ),
        );
        return result.toString();
      },
    );
  }

  /// Return a the lp token definition to use when creating a pool. Returns a JSON stringified
  /// [token1Symbol] is the symbol of the first token
  /// [token2Symbol] is the symbol of the second token
  Future<Result<String, Failure>> getLPTokenDefinition(
    String token1Symbol,
    String token2Symbol,
  ) async {
    return Result.guard(
      () async {
        final tokenDefinition = await apiService.callSCFunction(
          jsonRPCRequest: SCCallFunctionRequest(
            method: 'contract_fun',
            params: SCCallFunctionParams(
              contract: factoryAddress.toUpperCase(),
              function: 'get_lp_token_definition',
              args: [
                token1Symbol,
                token2Symbol,
              ],
            ),
          ),
        );
        return tokenDefinition.toString();
      },
    );
  }
}
